tests/tests.cpp

```diff
@@ -660,11 +660,49 @@ void test20() {
     cout << " done." << endl;
 }
 
+void test21() {
+    cout << "Running test 21...";
+    const char* txtFile = "this is some data";   // 17 Bytes
+    const char* txtFile2 = "this is some data!"; // 18 Bytes
+    int len = strlen(txtFile);
+    int len2 = strlen(txtFile2);
+
+    ZipArchive z1("test.zip");
+    z1.open(ZipArchive::WRITE);
+    z1.addData("somedata", txtFile, len);
+    z1.addData("somedata2", txtFile2, len2);
+    z1.close();
+
+    std::ifstream ifs("test.zip", std::ios::binary);
+    ifs.seekg(0, std::ifstream::end);
+    uint32_t bufferSize = (uint32_t)ifs.tellg();
+    char* buffer =  (char*)malloc(bufferSize);
```

Comment:
Missing `free(buffer)` at the end, probably causing the memory leak.

Reply:
It's not a leak, as @liweimax stated libzip deletes the buffer internally (confirmed by Valgrind too) !

I found the bug that generated the big Valgrind report above (btw sorry for pasting a long report) : we have only to remove these added lines, in  ZipArchive::close, to fix the memory bugs :

```
        if (zipSource){
//            zip_source_close(zipSource);
//            zip_source_free(zipSource);
//            zipSource = NULL;
        }
```

I don't have an explanation why...

However, this test rule is still failing, but at least, there's no huge Valgrind report like before :
`test_static: tests/tests.cpp:74: void test1(): Assertion `nbEntries==3' failed. `

Reply:
@liweimax Can you please remove the free(buffer) from test21(), it's not a memory leak and that free is causing an invalid free() detected by Valgrind (you said it yourself before : the buffer by freed internally).

Still, test_static is still failing : tests/tests.cpp:74: void test1(): Assertion `nbEntries==3' failed. I will investigate why the static version of tests is failing with this new feature when I find some free time.

Reply:
Ping @liweimax :sweat_smile: Is it possible to have an update about this ? Thanks !

Reply:
Done :-)